### ISSUE TITLE
Possible to generate multiple entry points into one output file?

### ISSUE THREAD

**Issue by pema99** [shader-slang/slang#3974]
Say I have a simple test shader, `test.slang`, containing a vertex and fragment entry point:

```hlsl
struct v2f
{
    int foobar;
}

int foo(int a)
{
    return a + 3;
}

[shader("vertex")]
v2f vert ()
{
    v2f res;
    res.foobar = foo(42);
    return res;
}

[shader("fragment")]
float4 frag (v2f i) : SV_Target
{
    return foo(i.foobar);
}
```

When compiling (to HLSL in this case), I must specify which entry point I wish to compile. This results in 2 output shaders (cut down a bit for brevity):

Vert:
```hlsl
int foo_0(int a_0)
{
    return a_0 + int(3);
}

struct v2f_0
{
    int foobar_0;
};

v2f_0 vert()
{
    v2f_0 res_0;
    res_0.foobar_0 = foo_0(int(42));
    return res_0;
}
```
Frag:
```hlsl
int foo_0(int a_0)
{
    return a_0 + int(3);
}

struct v2f_0
{
    int foobar_0;
};

float4 frag(v2f_0 i_0) : SV_TARGET
{
    return (float4)float(foo_0(i_0.foobar_0));
}
```

So, I have 2 questions about this setup:
- Is there any combination of options I can use to transpile both entry points into a _single_ HLSL file? Compiling entry points separately makes sense in most cases, but in my case, I'm feeding the resulting HLSL code forward into a separate analysis program, which expects a single file containing all entry points. In principle I could just concatenate the 2 resulting output files manually, but it would lead duplicate definitions of code shared between entry points (`foo` and `v2f`).
- If this isn't possible, is there a way to control the subtle name mangling Slang seems to do by default? The `foo` function and `v2f` struct have had `_0` appended to their name. If I could control this mangling scheme, and use a different number for the 2 output files, I could concatenate them without duplicate definitions.

I realize this is rather uncommon use case, and if it isn't feasible - so be it. I can think of a few ways of working around it, but they are quite a bit more involved, so I figured I would ask first.

**Comment by csyonghe** [comment]
Yes, we support multiple entrypoints when compiling to HLSL or SPIRV. You will need to specify multiple -entry arguments when using `slangc`, or link a program with more than one IEntryPoint when using the compilation API.

`slangc` example:
```
slangc test.slang -target hlsl -entry frag -profile ps_6_0 -entry vert -profile vs_6_0
```

**Comment by pema99** [comment]
> Yes, we support multiple entrypoints when compiling to HLSL or SPIRV. You will need to specify multiple -entry arguments when using `slangc`, or link a program with more than one IEntryPoint when using the compilation API.
> 
> `slangc` example:
> 
> ```
> slangc test.slang -target hlsl -entry frag -profile ps_6_0 -entry vert -profile vs_6_0
> ```

Isn't this command just letting you compile multiple entry points each to their own file? What I am asking for is the ability to take 1 single slang file with multiple entry points, and produce 1 single hlsl file with multiple entry points. Let me know if I've misunderstood you :)

**Comment by pema99** [comment]
If I run the command you posted, this output gets printed to stdout:
```
> ./slangc test.slang -target hlsl -entry frag -profile ps_6_0 -entry vert -profile vs_6_0 -line-directive-mode none
(0): warning 40: the 'sm_6_0' was specified more than once for target 'sm_6_0'
```
```hlsl
#pragma pack_matrix(column_major)
#ifdef SLANG_HLSL_ENABLE_NVAPI
#include "nvHLSLExtns.h"
#endif
#pragma warning(disable: 3557)

int foo_0(int a_0)
{
    return a_0 + int(3);
}

struct v2f_0
{
    int foobar_0;
};

float4 frag(v2f_0 i_0) : SV_TARGET
{
    return (float4)float(foo_0(i_0.foobar_0));
}

#pragma pack_matrix(column_major)
#ifdef SLANG_HLSL_ENABLE_NVAPI
#include "nvHLSLExtns.h"
#endif
#pragma warning(disable: 3557)

int foo_0(int a_0)
{
    return a_0 + int(3);
}

struct v2f_0
{
    int foobar_0;
};

v2f_0 vert()
{
    v2f_0 res_0;
    res_0.foobar_0 = foo_0(int(42));
    return res_0;
}
```
This just seems like 2 output files concatenated to me? If you treat this as 1 file, it has duplicate definitions.

If I try to write it to an output file, instead of stdout, I get this error (due to https://github.com/shader-slang/slang/issues/3973):
```
> ./slangc test.slang -target hlsl -entry frag -profile ps_6_0 -entry vert -profile vs_6_0 -line-directive-mode none -o test.hlsl
(0): warning 40: the 'sm_6_0' was specified more than once for target 'sm_6_0'
(0): error 70: the output path 'test.hlsl' is not associated with any entry point; a '-o' option for a compiled kernel must follow the '-entry' option for its corresponding entry point```

**Comment by csyonghe** [comment]
I took a look at the code and it appears that we simply don't support multiple entrypoints when using `slangc`.

This need to be done through the COM API.

**Comment by pema99** [comment]
> I took a look at the code and it appears that we simply don't support multiple entrypoints when using `slangc`.
> 
> This need to be done through the COM API.

Ok. Thanks for looking into it :) . I'll try to use the API. Treat this as a feature request, I guess.

**Comment by swoods-nv** [comment]
@pema99 can you let us know if using the API unblocks you? I'm going to add this to our Q4 backlog for the feature request in the meantime, but if you are blocked I can revisit.

**Comment by pema99** [comment]
Since the project I am working on is written entirely in managed code (C#), using the native API is a bit involved, and given the complexity of it, doesn't seem feasible to directly invoke from managed code. This is why I've been avoiding it thus far. Using `slangc` directly, and communicating via stdin/stdout made Slang extremely simple to integrate. If I go down the route of using the native API, I'll probably have to write a wrapper library which links again the slang library, and which is easier to call from managed code. That by itself isn't too bad, but it means I have to deal with the cross-platform problem myself, compiling the wrapper library for all platforms I wish to support. That's something I was hoping to avoid. I'm pretty close to having a working prototype of my project using `slangc`, which works fine so lang as there are no duplicate definitions; I can't think of any other blocking issues.

I just started fiddling with the native API today. I can let you know whether I manage to make it do what I am asking for in this thread. I'm following https://shader-slang.com/slang/user-guide/compiling.html#using-the-compilation-api right now, which doesn't really seem to make it clear how to put multiple entry points into a single output blob.

**Comment by swoods-nv** [comment]
@csyonghe - do we need to add to the API documentation here?

**Comment by pema99** [comment]
I found by trial and error that you can do what I want with `SLANG_TARGET_FLAG_GENERATE_WHOLE_PROGRAM` and the ICompileRequest API. Is this documented somewhere?

```c++
Slang::ComPtr<ICompileRequest> compileRequest(spCreateCompileRequest(globalSession));
compileRequest->setCodeGenTarget(SLANG_HLSL);
compileRequest->setTargetFlags(0, SLANG_TARGET_FLAG_GENERATE_WHOLE_PROGRAM);
int unitIdx = compileRequest->addTranslationUnit(SlangSourceLanguage::SLANG_SOURCE_LANGUAGE_SLANG, "mainUnit");
compileRequest->addTranslationUnitSourceFile(unitIdx, "test.slang");
SlangResult res = compileRequest->compile();
assert(SLANG_SUCCEEDED(res));

size_t blobSize;
const char* c = static_cast<const char*>(compileRequest->getCompileRequestCode(&blobSize));
printf("%s\n", c);
```

I couldn't find a way to do it with the ISession API, though I see there is a flag `CompilerOptionName::GenerateWholeProgram`, I can't figure out how the get the compiled code out at the end. `getEntryPointCode` doesn't do the trick.

**Comment by pema99** [comment]
Hmm... I discovered that you can ignore the COM API, and use the C-style API points `sp<name>()`. These are much simpler, and quite feasible to call directly from managed code. I'm unblocked by using this API :)

**Comment by csyonghe** [comment]
We are deprecating the old C-style API because they are also not flexible enough for complex use cases. Please see our documentation for the COM API here:
https://shader-slang.com/slang/user-guide/compiling.html

**Comment by swoods-nv** [comment]
@csyonghe - If there's stuff you can do for simple use cases with the C-style API that you can't with the COM API, that seems like it might be a gap we should address in the COM API.

**Comment by pema99** [comment]
> @csyonghe - If there's stuff you can do for simple use cases with the C-style API that you can't with the COM API, that seems like it might be a gap we should address in the COM API.

I'm sure the API is plenty powerful. It's not that the C API can do something the COM API cannot - my issue is that the signatures in the COM API are much more complex. They don't really seem like something you'd want to use outside of a C++ codebase to me. The C-style API seems much easier to use from a language that isn't C++, such as C# in my case.

Just to illustrate, here are some example bindings I've just written:
```cs
[DllImport("slang", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true)]
private static extern IntPtr spCreateCompileRequest(IntPtr session);

[DllImport("slang", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true)]
private static extern void spDestroyCompileRequest(IntPtr request);
```

The COM API equivalents would involve marshalling rather complex looking objects with vtables and what not back and forth. That kind of thing is usually pretty unpleasant in usecases like mine.

**Comment by pema99** [comment]
Apropos, is the `ICompileRequest` object deprecated alongside the C-API? I don't see a way to create such a request outside of `spCreateCompileRequest`.

I'll probably use the C-style API for now. If `slangc` gains the ability to do whole-program compilation at some point, I might switch back to that. If the C-style API gets completely removed in the future, I guess I'll take another look at making C# bindings against the COM API. My first attempts didn't achieve much more than making my head spin.

**Comment by swoods-nv** [comment]
Ok-- I'm going to mark this as a future need, because it's a rough corner that should be addressed, but you have a way forward currently.

**Comment by pema99** [comment]
Thanks to both of you for all the help :)

**Comment by pborsutzki** [comment]
I'd also be interested in generating multiple entry points into one output file - specifically for the C++ and the SPIRV backends from the `slangc` command.

I'd also wish for simply putting all entry points of the given slang source file into the output, without explicit having to specify them via `-entry` arguments. I'd hope marking functions with `[shader("...")]` attributes should be sufficient to mark entry points?

**Comment by csyonghe** [comment]
@pborsutzki Our spirv backend already allow multi entry points. You can do:

slangc test.slang -target spirv -emit-spirv-directly -o out.spv

The output will include all entry points defined in the source as long as they are decorated with [shader] attribute.

**Comment by csyonghe** [comment]
For C++, I believe this is also supported today via API.

**Comment by csyonghe** [comment]
@pborsutzki I am closing this issue as it seems you are able to get desired behavior. @pema99 I agree that given the current state using COM API directly from C# is difficult. I think the long term solution required here is for somebody to create a C# wrapper from all the COM interface types we exposed.

**Comment by pema99** [comment]
Yeah, probably so. I might give it another attempt at some point. I'm not sure how to marshal COM objects, and how to deal with the reference counting and the vtables for virtual methods, but I'm sure there's some way to do it.

**Comment by csyonghe** [comment]
This documentation seems relavant:
https://learn.microsoft.com/en-us/dotnet/standard/native-interop/comwrappers-source-generation

**Comment by mmgeorge** [comment]
Is this also supported for wgsl? Looking at the [example](https://github.com/shader-slang/slang/tree/master/examples/wgpu-html5) If I do: 
```
slangc -target wgsl -o shader.wgsl .\shader.slang
```
I'm getting: 
```
error 70: the output path 'shader.wgsl' is not associated with any entry point; a '-o' option for
a compiled kernel must follow the '-entry' option for its corresponding entry point
```

And I'm not seeing any additional args I can pass when looking at the [guide](https://shader-slang.org/slang/user-guide/wgsl-target-specific.html). Thanks!

**Comment by csyonghe** [comment]
This might be an oversight. Can you file a separate issue on this?
